### ISSUE TITLE
Debugging a stackalloc line takes multiple F10s

### ISSUE THREAD

**Issue by eerhardt** [dotnet/coreclr#14926]
I’ve been debugging some code that looks like:

```C#
float *ptr = stackalloc float[10];

Span<float> span1 = stackalloc float[20];
span1[0] = 9.5f;
span1[1] = 8.5f;
span1[2] = 7.5f;
span1[3] = 6.5f;

Span<float> span2 = stackalloc float[4];
span2[0] = .5f;
span2[1] = 10.5f;
span2[2] = 10.5f;
span2[3] = 10.5f;
```

When debugging, it is taking multiple “F10”s in the debugger to get over the stackalloc lines.  I’ve noticed that it grows as the more memory is alloc’d.  But even in the small case (alloc’ing 1 element), it still takes roughly 5 times of hitting “F10” to get past the line.

I’m using:
- .Net Core 2.1.0-preview1-25907-02
- Microsoft Visual Studio Enterprise 2017 Int Preview (2)
  - Version 15.5.0 Preview 4.0 [27106.0.d15rel]
  - VisualStudio.15.IntPreview/15.5.0-pre.4.0+27106.0.d15rel

/cc @AndyAyersMS @mikem8361 @VSadov @stephentoub

**Comment by AndyAyersMS** [comment]
I'll take a look.

**Comment by AndyAyersMS** [comment]
Think the issue may be on the debugger side. The jit's IL to native map looks reasonable:

```
IL offs 0x0000 : 0x00000041 ( STACK_EMPTY )
IL offs 0x0001 : 0x00000042 ( STACK_EMPTY )
IL offs 0x0004 : 0x00000049 ( STACK_EMPTY )
IL offs 0x000B : 0x00000094 ( CALL_INSTRUCTION )
```
and when debugging we seem to get hung up in the code that is zeroing the stackalloc:
```asm
000073 push     0
000075 push     0
000077 dec      rdx
00007A jne      SHORT G_M60923_IG05
```
Maybe the fact that RSP is moving and temporarily unaligned in this code stretch is causing issues?

When I debug this under VS2017 (preview 4 27109.1) and go to "view disassembly" I get some assert popups:

![image](https://user-images.githubusercontent.com/10121823/32743294-d6cbd444-c860-11e7-80a0-c7374823a94f.png)

If I ignore this and a second assert the disassembly window takes a long time to populate but when it does it looks reasonable:
![image](https://user-images.githubusercontent.com/10121823/32743736-1fec765a-c862-11e7-9a6d-081fd189489d.png)

Source-level F10s at this point cycle through the loop that zeros the stackalloc.

**Comment by lt72** [comment]
We will move this back to 2.1 if we have time. Currently this does not fit in our planning.

**Comment by cshung** [comment]
# Overview
The root cause of the issue is that the stepper would like to make sure, when we are stepping over a line, we are on the same frame, but the changing `rsp` is breaking the wrong assumption in the debugger that `rsp` is an unique identifier for the frame. Changing that assumption is non-trivial.

# Here is the analysis:

In our scenario, if [DebuggerStepper::TriggerSingleStep(...)](https://github.com/dotnet/coreclr/blob/df0cd2cd2c026af3aff78206bed6b45429dc9e19/src/debug/ee/controller.cpp#L7382) returns true, the debugger would believe the stepping completes.

As one of the determining whether or not the stepping should stop is the [DebuggerStepper::IsInRange(...)](https://github.com/dotnet/coreclr/blob/df0cd2cd2c026af3aff78206bed6b45429dc9e19/src/debug/ee/controller.cpp#L5196) check. If we are still in the range, we should continue stepping. Since we are stepping just a single line without any function call, I expect we should be considered in range when we hit the `push 0` instruction, but that's not true as the code runs.

The `IsInRange(...)` check fails because the range is not appropriate according to [DebuggerStepper::IsRangeAppropriate(...)](https://github.com/dotnet/coreclr/blob/df0cd2cd2c026af3aff78206bed6b45429dc9e19/src/debug/ee/controller.cpp#L5108) check. In particular, the function believe the "frame pointer" is not the same as the "frame pointer" captured when the stepper is constructed (i.e. `info->m_activeFrame.fp != m_fp).

Looking at the assembly code, the `rbp` is not changed at all. (Of course, except the only changes in prolog and epilog as expected). Here is the interesting part, the [FramePointer](https://github.com/dotnet/coreclr/blob/78a5ee8b39fc5664b47dea447ed2fb00d2c36afa/src/debug/inc/dbgipcevents.h#L1249) struct actually store `rsp`.

So now it is clear - of course, the `rsp` changes.

# Next steps:
The obvious thing to do is the correct the assumption. On any architecture that allows `stackalloc`, stack pointer changes. On ARM/ARM64, it is possible for the leaf frame to have the same stack pointer (as return address is stored on the link register). 

However, the struct `FramePointer` is used extensively throughout the debugger code base with 278 occurrences, changing it wouldn't be easy.

@dotnet/dotnet-diag

**Comment by cshung** [comment]
For your information @BruceForstall .

**Comment by sdmaclea** [comment]
@noahfalk @mikem8361 

This was the PR I was commenting on in the stand up a few days ago.

Per @cshung analysis above, the https://github.com/dotnet/coreclr/blob/df0cd2cd2c026af3aff78206bed6b45429dc9e19/src/debug/ee/controller.cpp#L5108 check is using the StackPointer instead of the FramePointer in this code.

https://github.com/dotnet/coreclr/blob/df0cd2cd2c026af3aff78206bed6b45429dc9e19/src/debug/ee/controller.cpp#L5135-L5141

**Comment by noahfalk** [comment]
I'd be supportive of fixing this is in 5.0, but I don't see a path where we could get enough confidence in the fix to ship it in 3.0. My impression of the stack walking code is that it has many layers of abstractions and various warts that have accumulated over a long time which makes analyzing it time consuming and error-prone. If we are going to make an attempt I'd recommend doing it towards the beginning of a release so we have plenty of time to find and fix any regressions we cause.


My best guess at some broad strokes to fix it would be:
a) Rename FramePointer to StackPointer to be more transparent about what the current code is doing
b) Create a new FramePointer type that is more true to its name and does not vary when stackalloc is used. We'd probably have to be fairly explicit in the definition because there are a number of things that can add confusion:
  - Not all methods set the architecture frame pointer (rbp I think is mostly unused)
  - Some methods use explicit clr!Frame structures homed somewhere on the stack
  - On some architectures we use funclets which may or may not be treated as being a unified frame with their parent depending on the context
c) Hopefully find a scope of code where we can use FramePointer instead of StackPointer that is broad enough to address the stepping issue while still being narrow enough that we can code review it decently.

I'm not sure how much that helped since there remains a good bit ambiguity. I could probably make a narrower suggestion if needed, but I'd need to take more time for code review first.

**Comment by sdmaclea** [comment]
This is fixed in the 5.0 branch.
